File: challenges/javascript/search-book.js
function authorNames(catalogData, book) {
  return _.map(_.get(book, "authorIds"),
               function(authorId) {
    return _.get(catalogData, ["authorsById", authorId, "name"]);
  });
}

function bookInfo(catalogData, book) {
  return  {
    "title": _.get(book, "title"),
    "isbn": _.get(book, "isbn"),
    "authorNames": authorNames(catalogData, book)
  };
}

function searchBooksByTitle(libraryData, query) {
  var catalogData = _.get(libraryData, "catalog");
  var allBooks = _.get(catalogData, "booksByIsbn");
  var matchingBooks = _.filter(allBooks, function(book) { 
    return _.get(book, "title").toLowerCase()
      .includes(query.toLowerCase());
  });
  return JSON.stringify(_.map(matchingBooks, function(book) {
    return bookInfo(catalogData, book);
  }));
}

searchBooksByTitle(libraryData, "watCH");

File: challenges/ruby/search-book.rb
load 'library-data.rb'

def author_names(catalog_data, book)
  book['authorIds'].map do |author_id|
    catalog_data.dig('authorsById', author_id, 'name')
  end
end

def book_info(catalog_data, book)
  {
    'title'       => book['title'],
    'isbn'        => book['isbn'],
    'authorNames' => author_names(catalog_data, book)
  }
end

def search_books_by_title(library_data, query)
  catalog_data   = library_data['catalog']
  all_books      = catalog_data['booksByIsbn']
  matching_books = all_books.select do |_, book|
    book['title'].downcase.include?(query.downcase)
  end
  matching_books.map { |_, book| book_info(catalog_data, book) }.to_json
end

pp search_books_by_title($library_data, 'watCH')
